### ISSUE TITLE
matrix returned by numpy.bmat() does not work as expected with @ operator

### ISSUE THREAD

**Issue by fschaffalitzky** [numpy/numpy#8712]
[bmatbug.py.txt](https://github.com/numpy/numpy/files/805744/bmatbug.py.txt)

Using numpy.bmat(...) to generate a 2-D array produces a result that, when multiplied with a 1-D vector using the @ operator, does not behave the same way as a 2-D array of the same size produced by calling numpy.copy(...) on the original one.  Illustrative pseudo-code:

```
x = numpy.random.random((3,))        # x.shape is (3,)
a = numpy.bmat([...])                # a.shape is (5,3)
y = a @ x                            # y.shape is (1,5)   *** wrong? ***
y = numpy.copy(a) @ x                # y.shape is (5,)
```
What is also curious is that both 'a' and numpy.copy(a) have the exact same .shape and .strides.

The attached script reproduces the behavior, generating the following output:

```
$ python3 -B bmatbug.py 
x (3,) (8,)
---
a (5, 3) (24, 8)
y (1, 5) (40, 8)
---
a (5, 3) (24, 8)
y (5,) (8,)
```

--- system parameters ---

Python version: 3.5.2
numpy version: 1.11.0 and 1.11.2
OS: Ubuntu and macOS
Output of 'uname -a':
  "Linux xxx 4.4.0-64-generic #85-Ubuntu SMP Mon Feb 20 11:50:30 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux" (Ubuntu 16.10)
  "Darwin xxx 16.4.0 Darwin Kernel Version 16.4.0: Thu Dec 22 22:53:21 PST 2016; root:xnu-3789.41.3~3/RELEASE_X86_64 x86_64" (macOS Sierra 10.12.3)

**Comment by eric-wieser** [comment]
`np.bmat` returns an `np.matrix`, which has the frustrating behaviour of refusing to become 1d. This type is contagious, and takes priority in any operation with a raw array

**Comment by fschaffalitzky** [comment]
OK, now it all makes sense.  I would still argue that the operation "matrix @ ndarray" returning a 1 x N matrix is an odd behavior, since matrices times anything don't do that unless they have only one row, and this should have been a runtime error instead, but your explanation was very helpful since I realized I can use bmat(...).A instead of copy(bmat(...)).  Thanks for your attention and help.

**Comment by eric-wieser** [comment]
> since matrices times anything don't do that unless they have only one row

What do you mean by this?

**Comment by fschaffalitzky** [comment]
Sorry, I didn't express that very well.

What I meant is that, at least outside of numpy, the traditional concept of matrices has the property that the C = A * B has the same number of rows as A and the same number of columns as B.  So it was surprising to me that a class called "matrix" whose design objective appears to be to behave more like a matrix and less like an ndarray would implement A @ B in a way where the result has 1 row and 5 columns (i.e. is 1 x 5) even when A has 5 rows and 3 columns (i.e. is 5 x 3).  If the objective of numpy.matrix was to reproduce matrix-like behavior, I would consider this implementation a failure.  Or, if I used the class in a context that was not intended, it would have been more helpful to get a run-time/type error.

That's just griping, of course.  I assume that the interface of numpy.matrix is traditional and changing it at this point might be more trouble than it is worth.

**Comment by seberg** [comment]
I would argue that this actually should somewhat be fixed by implementing `__matmul__` in `matrix`. It is probably not a high priority considering that the main prupose of having `@` in *python* even is to get rid of the matrix class. So while I would say that fixing this would not be bad, probably most of us don't really care about it, since we consider matrix to be outdated.

**Comment by eric-wieser** [comment]
Sorry, I see the bug now. Your complaint is that the shape is `(1,5)` and not `(5,1)`

**Comment by seberg** [comment]
Yes, it is wrong, and while some of us may not care much, it would be better if it gets fixed.

**Comment by fschaffalitzky** [comment]
Sorry for being confusing at first.  Personally I would prefer if a new function numpy.blockmatrix(...) was added that behaves like numpy.bmat(...) but which returns numpy.ndarray instead of numpy.matrix

**Comment by NeilGirdhar** [comment]
@fschaffalitzky I just came here to suggest exactly this after I got bitten by [the same thing](http://stackoverflow.com/questions/43176131/how-do-numpy-block-matrices-work)! Those `matrix` classes have become poisonous traps.
